### PR TITLE
[Table Visualization] Replace div containers with OuiFlex components

### PR DESCRIPTION
**Pull request by curq**
### Description

Replaces the usage of div container with OuiFlex components for better OuiCompliance. Reduced custom styling and removed a custom class.

Explanation why margin-bottom and padding where removed:
1. ```margin-bottom: $euiSizeL;``` is replaced by gutterSize prop (which adds negative margin to add distance between flex items, default value 'l' (-12px)). So there is no longer need to add margin-bottom.
2. ```padding: $euiSizeS;``` was also somewhat replaced by gutterSize.

---

Also, it looks like switching to Flex components changed the column view.
**before:**
beforeColumnView
**after:**
afterColumnView

I think the after view is a better option from visual perspective so there is no need to try keeping before view. What are you thoughts?

---

Another important change is that in row view the tables are now distributed equally through the entire table container instead of taking all space that it needs. Here how it was before:
5 items per page
image
30 items per page
image

This is how it is now:
5 items per page
image

30 items per page
image

It only happens if we change https://github.com/opensearch-project/OpenSearch-Dashboards/blob/55b293a2e52399c5c4101f5b3e643519c4531809/src/plugins/vis_type_table/public/components/table_vis_component_group.tsx#L25-L34 to OuiFlexItem
I tried investigating deeper, but I didn't find the reason for such behavior change. Maybe it is related how OuiDataGrid is implemented?
I'm not sure which behavior is expected, but I think the 'after' one is more user friendly?

### Issues Resolved
Solves https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4255

## Screenshot
To provide clear view how looks changed here are before and after screenshots for both column and row view
### Before
row view:
beforeRowView

column view:
beforeColumnView

### After
row view:
afterRowView

column view:
afterColumnView

## Testing the changes
Use split tables option and then check both column and row views, it is also good to check how it behaves with large number of rows.

### Check List

- [ ] All tests pass
  - [x] `yarn test:jest`
  - [x] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by joshuarrrr** [comment]
> I think the after view is a better option from visual perspective so there is no need to try keeping before view. What are you thoughts?

I agree that it seems better. But when testing and validating updates to visualizations, it's important to also see how the behavior changes when the visualization is embedded in a dashboard (as well as on the visualization page, as the screenshots here cover).

> Another important change is that in row view the tables are now distributed equally through the entire table container instead of taking all space that it needs.

I also like the after behavior, (pending validation in dashboard mode with small panels). But in the Screenshot -> After -> row view screenshot, the behavior looks a little different - we may need to adjust so that pagination controls always appear directly beneath the last row, not all the way at the bottom of a stretched flex item.

I'll pull this PR down and do some additional validation and screenshots.

File: src/plugins/vis_type_table/public/components/table_vis_app.scss
Comment (curq) [reply src/plugins/vis_type_table/public/components/table_vis_app.scss:4]: We could wait for this one to finish https://github.com/opensearch-project/oui/issues/776, and this one for utility ( we don't have option for overflow:auto right now ) https://github.com/opensearch-project/oui/issues/774.

File: src/plugins/vis_type_table/public/components/table_vis_app.scss
Comment (curq) [reply src/plugins/vis_type_table/public/components/table_vis_app.scss:11]: We have 2 options right now.
1. If https://github.com/opensearch-project/oui/issues/776 will be accepted and PR made we could just move it to component property.
2. I tried to remove them and noticed that this styling has no effect on layout at all. I think it needs more testing and looking around so we can make sure that it really can be removed.

File: src/plugins/vis_type_table/public/components/table_vis_app.test.tsx
Comment (curq) [reply src/plugins/vis_type_table/public/components/table_vis_app.test.tsx:78]: So we can remove it, right? Also, should this one be also removed? https://github.com/opensearch-project/OpenSearch-Dashboards/blob/55b293a2e52399c5c4101f5b3e643519c4531809/src/plugins/vis_type_table/public/components/table_vis_app.test.tsx#L82-L98
I will try to think about how we can test it, but for now it looks like after we converted to OUI it is not required as direction is handled by OUI components now.

**Comment by curq** [comment]
@joshuarrrr, I tried removing rest of the styles and it looks like layout didn't change at all. I'm still not sure if it safe to remove them, as before changing to Flex components removing these styles did break the layout. As I understand the ```OuiFlex``` and ```OuiFlexGroup``` do apply shrink and basis by default, although I'm not sure if it the exact values as in the current custom stylings. So, I have suspicion that FlexItem and FlexGroup are already handle our layout. But like I said, we need to carefully look through the layout.

**Comment by ananzh** [comment]
@curq I like the change. But I see a regression after quick check.

Before 

Screenshot 2023-06-09 at 14 55 43

After
Screenshot 2023-06-09 at 14 55 43

See the spaces are missing on the left side of table. This is just table with a large time range. No split. We need to make sure the change won't have any regression issues. Could you check if you see this issue?

**Comment by curq** [comment]
@ananzh Thanks for the catch! The issue with missing space was due to ```TableVisComponent``` (used when only single table is rendered) was wrapped in ```</>``` empty tag, instead of ```FlexItem``` like in ```TableVisComponentGroup```, so the property ```gutterSize``` of FlexGroup was not applying to it, resulting in the lack of spacing. I fixed it and now it should be fine. Could you please take a look?
image

**Comment by ananzh** [comment]
> @ananzh Thanks for the catch! The issue with missing space was due to `TableVisComponent` (used when only single table is rendered) was wrapped in `</>` empty tag, instead of `FlexItem` like in `TableVisComponentGroup`, so the property `gutterSize` of FlexGroup was not applying to it, resulting in the lack of spacing. I fixed it and now it should be fine. Could you please take a look? image

Nice!! Thank you~

**Review by ananzh — DISMISSED** [review]

File: src/plugins/vis_type_table/public/components/table_vis_app.scss
Comment (ananzh) [reply src/plugins/vis_type_table/public/components/table_vis_app.scss:11]: Curious, @curq, could we do `<EuiFlexItem grow={false}>` here? I think the `grow` prop of EuiFlexItem behaves similarly to the 'flex-grow' CSS property. So by default, EuiFlexItem components will grow and shrink as needed (i.e., flex: 1 1 auto). If you want to mimic the behavior of flex: 0 0 auto;, you should set the 'grow' property of the EuiFlexItem to false...

Not a blocker for this PR.

Comment (joshuarrrr) [reply src/plugins/vis_type_table/public/components/table_vis_app.scss:11]: @curq Let's just make a follow-up issue for this and we can tackle it after this PR.

**Review by joshuarrrr — DISMISSED** [review]
Approved, pending successful tests.

**Comment by joshuarrrr** [comment]
The failing cypress test is:

```
  1 failing

  1) Vis Builder: Table Chart
       Basic test:
     AssertionError: Timed out retrying after 60000ms: Expected to find element: `[class="visTable__group"]`, but never found it. Queried from element: <div.euiFlexGroup.euiFlexGroup--gutterLarge.euiFlexGroup--directionColumn.euiFlexGroup--responsive.visTable>
      at testSplitTables (http://localhost:5601/__cypress/tests?p=cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/vis_builder/vis_types/table.spec.js:170:118)
      at Context.eval (http://localhost:5601/__cypress/tests?p=cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/vis_builder/vis_types/table.spec.js:134:7)
```

This PR leaves that class intact, so I'd expect it to pass. @ananzh Can you potentially take a look at that test definition and help @curq figure out if there's a regression or if the test need updating? https://github.com/opensearch-project/opensearch-dashboards-functional-test/blob/6823c8202eb8aca8ec1ef1dee41c9e25f65622ab/cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/vis_builder/vis_types/table.spec.js#L88-L96

**Comment by curq** [comment]
@joshuarrrr I think this fail is caused by this line. https://github.com/opensearch-project/opensearch-dashboards-functional-test/blob/6823c8202eb8aca8ec1ef1dee41c9e25f65622ab/cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/vis_builder/vis_types/table.spec.js#L60
```
export const testSplitTables = (dir, num) => {
  cy.getElementByTestId('visTable')
    .should('have.class', `visTable ${dir}`.trim())
    .find('[class="visTable__group"]')
    .should(($tables) => {
      // should have found specified number of tables
      expect($tables).to.have.length(num);
    });
};
```
the test 
```
testSplitTables('visTable__groupInColumns', 4);
```
tries to find element with `visTable__groupInColumns`, but it was removed in this PR after we moved the direction logic to OUI components. So the test fails. I believe we need to update it.

**Comment by joshuarrrr** [comment]
https://github.com/opensearch-project/OpenSearch-Dashboards/assets/1679762/01e6f359-12b2-48c7-a6ad-b5cd57a28f57

**Comment by joshuarrrr** [comment]
Vis Builder Table Chart -- Basic test (failed)

**Comment by ananzh** [comment]
@curq @joshuarrrr yes, test failed due to `visTable__groupInColumns` removed. 
@joshuarrrr should we ask @curq to fix the test as a completion of this PR or should I work on it since understand and setup cypress also need some work? Not sure if @curq has time to work on this.

**Comment by joshuarrrr** [comment]
@ananzh - @curq has opened https://github.com/opensearch-project/opensearch-dashboards-functional-test/issues/730 and https://github.com/opensearch-project/opensearch-dashboards-functional-test/pull/731. Let's first review and approve this PR, but hold merging until we also review and approve https://github.com/opensearch-project/opensearch-dashboards-functional-test/pull/731. That way we can merge this, validate the tests pass in the functional test repo, then merge the test update.

**Comment by joshuarrrr** [comment]
@curq I'm not sure all the functional tests are completely updated? It looks like the class is still present here: https://github.com/opensearch-project/opensearch-dashboards-functional-test/blob/7f8a7d799abe560549759fa440b0184a3238b61b/cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/vis_type_table/split.spec.js#L314

Can you try running the cypress tests against this branch again? The failure we're seeing is actually looking for the `visTable__group` string.

To run the cypress tests against your local OpenSearch Dashboards, start it with `yarn start --no-base-path` (to use port `5601` directly), and make sure both `5601` and `9200` ports are accessible, then run `yarn cypress:open` from the functional test repo, and you can choose a particular spec to run and debug.

**Comment by curq** [comment]
@joshuarrrr Thanks for the feedback! You're right it looks like there is single usage of `visTable__groupInColumns` left at the file you mentioned. Test should be fine after we remove it. I rechecked the codebase and it is the only usage of this class left. 

I made a quick fix PR. Could check it please? https://github.com/opensearch-project/opensearch-dashboards-functional-test/pull/785

**Comment by curq** [comment]
> Can you try running the cypress tests against this branch again? The failure we're seeing is actually looking for the `visTable__group` string.

> To run the cypress tests against your local OpenSearch Dashboards, start it with `yarn start --no-base-path` (to use port `5601` directly), and make sure both `5601` and `9200` ports are accessible, then run `yarn cypress:open` from the functional test repo, and you can choose a particular spec to run and debug.

@joshuarrrr I didn't manage to run these in my previous fix so I wasn't able to fully check solution (I use docker on mac m1). I checked the CI and noticed that it still fails. 

Today I made it work so after looking through the errors, it looks like the line  `.find('[class="visTable__group"]')` https://github.com/opensearch-project/opensearch-dashboards-functional-test/blob/7f8a7d799abe560549759fa440b0184a3238b61b/cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/vis_builder/vis_types/table.spec.js#L91
Was causing the issue, because it matches elements with exactly single class `visTable__group`, but because now the container div also got `euiFlexItem` class, it failed to match.

I made [another PR](https://github.com/opensearch-project/opensearch-dashboards-functional-test/pull/787) that should finally make the tests pass. I checked it on my local dashboards instance and this error is indeed fixed. In the given test update PR I provided screenshot of successful test pass. There were other tests error but I think they were unrelated to this PR.

**Comment by joshuarrrr** [comment]
all tests pass!

**Review by ananzh — APPROVED** [review]

**Review by joshuarrrr — APPROVED** [review]

**Pull request merged**
